README.md
# Sharetribe Flex Template for Web: time-based process

[![CircleCI](https://circleci.com/gh/triple-point-organization/culinary-chefz/tree/master.svg?style=svg&circle-token=958e292bcede5b399299f83f36ad4730f908ec90)](https://circleci.com/gh/triple-point-organization/culinary-chefz/tree/master)

> **Note** If you are using day-based booking process (i.e. daily or nightly bookings), you should
> choose the original [Flex Template for Web](https://github.com/sharetribe/ftw-daily/) instead.

This is a template web application for a Sharetribe Flex marketplace using **time-based booking
process**. This repository is a fork from the existing
[FTW repository](https://github.com/sharetribe/ftw-hourly) (which is using nightly booking process).
So, if you have started working with the original FTW repository and you want to change to
FTW-hourly, you can
[change the 'upstream' remote repository](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/configuring-a-remote-for-a-fork).
You can read more about time-based process from the related
[Flex Docs article](https://www.sharetribe.com/docs/background/time-based-template)

## Quick start

> **Note:** this template assumes that you are using
> [time-based transaction process](https://www.sharetribe.com/docs/guides/how-to-take-time-based-bookings-into-use/)
> with alias `flex-hourly-default-process/release-1`, if you are using something else, you should
> change the alias name from
> [config.js](https://github.com/sharetribe/ftw-hourly/blob/master/src/config.js#L35) and
> [util/transaction.js](https://github.com/sharetribe/ftw-hourly/blob/master/src/util/transaction.js#L111)

If you just want to get the app running quickly to test it out, first install
[Node.js](https://nodejs.org/) and [Yarn](https://yarnpkg.com/), and follow along:

```sh
git clone git@github.com:sharetribe/ftw-hourly.git             # clone this repository
cd ftw-hourly/                                                 # change to the cloned directory
yarn install                                                   # install dependencies
yarn run config                                                # add the mandatory env vars to your local config
yarn run dev                                                   # start the dev server, this will open a browser in localhost:3000
```

You can also follow along the
[Getting started with FTW](https://www.sharetribe.com/docs/tutorials/getting-started-with-ftw/)
tutorial and read the background article about
[Time-based template](https://www.sharetribe.com/docs/background/time-based-template/) from the
[Flex Docs website](https://www.sharetribe.com/docs/).

For more information of the configuration, see the
[FTW Environment configuration variables](https://www.sharetribe.com/docs/references/ftw-env/)
reference in Flex Docs.

> **Note:** If you want to build your own Flex marketplace on top of the template, you should fork
> the repository instead of cloning it. See the
> [How to Customize FTW](https://www.sharetribe.com/docs/guides/how-to-customize-ftw/) guide in Flex
> Docs.

### For Windows users

Change `export` to `set` in the package.json file if you're using Windows/DOS. You need to do the
change to "dev" and "dev-sever" commands.

```
"dev": "yarn run config-check&&set NODE_ENV=development REACT_APP_DEV_API_SERVER_PORT=3500&&concurrently --kill-others \"yarn run dev-frontend\" \"yarn run dev-backend\""
```

```
"dev-server": "set NODE_ENV=development PORT=4000 REACT_APP_CANONICAL_ROOT_URL=http://localhost:4000&&yarn run build&&nodemon --watch server server/index.js"
```

We strongly recommend installing
[Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/about), if you are
developing on Windows. These templates are made for Unix-like web services which is the most common
environment type on host-services for web apps. Also, Flex Docs uses Unix-like commands in articles
instead of DOS commands.

## Getting started with your own customization

If you want to build your own Flex marketplace by customizing the template application, see the
[How to Customize FTW](https://www.sharetribe.com/docs/guides/how-to-customize-ftw/) guide in Flex
Docs.

## Deploying to Heroku

> **Note:** Remember to fork the repository before deploying the application. Connecting your own
> Github repository to Heroku will make manual deploys easier.

See the
[How to deploy FTW to production](https://www.sharetribe.com/docs/guides/how-to-deploy-ftw-to-production/)
guide in Flex Docs for more information.

[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy)

## Documentation

See the Flex Docs site: https://www.sharetribe.com/docs/

See also the [docs/](docs/) directory for some additional internal documentation.

## Get help – join Sharetribe Flex Developer Slack channel

If you have any questions about development, the best place to ask them is the Flex Developer Slack
channel at https://www.sharetribe.com/flex-slack

## License

This project is licensed under the terms of the Apache-2.0 license.

See [LICENSE](LICENSE)


CHANGELOG.md
# Change Log

We are not following semantic versioning in this template app since any change could potentially be
a breaking change for forked customization projects. We are still experimenting with what is a good
way to update this template, but currently, we follow a pattern:

- Major version change (v**X**.0.0): Changes to several pages and other components. Consider
  implementing this without merging upstream (we'll provide instructions).
- Minor version change (v0.**X**.0): New features and changes to a single page. These are likely to
  cause conflicts.
- Patch (v0.0.**X**): Bug fixes and small changes to components.

Note: Previous release tags (before v4.0.0) are coming from upstream:
https://github.com/sharetribe/flex-template-web/

---

## Upcoming version 2020-XX-XX

- [fix] Handle changing the start time in BookingTimeForm so that it doesn't cause error for
  fetching the lineItems. [#143](https://github.com/sharetribe/ftw-hourly/pull/143)

## [v10.1.1] 2021-04-20

- [fix] currency for Poland (PLN) [#140](https://github.com/sharetribe/ftw-hourly/pull/140)

### Updates from upstream (FTW-daily v8.1.0)

- [change] Update jose to v3.11.4 [#1433](https://github.com/sharetribe/ftw-daily/pull/1433)
- [add] Update fr.json, es.json and partially de.json
  [#1431](https://github.com/sharetribe/ftw-daily/pull/1431)
- [fix] currency conversion should not expect that env-variable is set.
  [#1425](https://github.com/sharetribe/ftw-daily/pull/1425)
- [fix] LoadableComponentErrorBoundary should be used in prod, not in dev-mode with
  hot-loading.[#1429](https://github.com/sharetribe/ftw-daily/pull/1429)
- [fix] currency for Poland (PLN) [#1427](https://github.com/sharetribe/ftw-daily/pull/1427)

  [v10.1.1]: https://github.com/sharetribe/ftw-hourly/compare/v10.1.0...v10.1.1

## [v10.1.0] 2021-03-11

### Updates from upstream (FTW-daily v8.1.0)

- [change] Specify required Node.js versions in package.json and update the node version used in
  CircleCI. Currently, the required Node.js version comes from
  [jose](https://github.com/panva/jose#runtime-support-matrix) package which is used with social
  logins. [#1418](https://github.com/sharetribe/ftw-daily/pull/1418)
- [fix] enforce upper case for currency and improve error message for it.
  [#1417](https://github.com/sharetribe/ftw-daily/pull/1417)
- [add] Add `LoadableComponentErrorBoundary` for handling ChunkLoadErrors with error boundary.
  [#1416](https://github.com/sharetribe/ftw-daily/pull/1416)

  [v10.1.0]: https://github.com/sharetribe/ftw-hourly/compare/v10.0.0...v10.1.0

## [v10.0.0] 2021-02-17

This major release adds support for code-splitting using
[Loadable Components](https://loadable-components.com/). At this point, we added route-based code
splitting, which meant changes to routeConfiguration.js and how "loadData" & "setInitialValues"
functions are defined and passed to routeConfiguration. Read more from
[Flex Docs](https://www.sharetribe.com/docs/ftw-routing/how-code-splitting-works-in-ftw/) and
related pull requests.

### Updates from upstream (FTW-daily v8.0.0)

- [fix] Remove unintended Lodash usage, unspecified window-scope calls and unused vars
  [#1413](https://github.com/sharetribe/ftw-daily/pull/1413)
- [add] Route-based code splitting. This is done against sharetribe-scripts v5.0.0 using Loadable
  components. Read more from the pull request.
  [#1411](https://github.com/sharetribe/ftw-daily/pull/1411)

  [v10.0.0]: https://github.com/sharetribe/ftw-hourly/compare/v9.3.0...v10.0.0

## [v9.3.0] 2021-01-13

- [fix] Use dayCountAvailableForBooking config instead of hard-coded value in FieldDateAndTimeInput
  so that it matches the date range that is used in react-dates components.
  [#129](https://github.com/sharetribe/ftw-hourly/pull/129)
- [fix] Use default timezone when fetching availability exceptions if availabilityPlan and
  information about listing's timezone doesn't exist yet.
  [#130](https://github.com/sharetribe/ftw-hourly/pull/130)

### Updates from upstream (FTW-daily v7.3.0)

- [fix] Move well-known/\* endpoints related to OIDC proxy setup from `apiRouter` to new
  `wellKnownRouter`so that they can be enabled outside the basic auth setup. It also makes it
  simpler to set the identity provider url, because we can drop the `/api` part of the path. Also,
  rename the `RSA_SECRET_KEY`to `RSA_PRIVATE_KEY` for consistency.
  [#1399](https://github.com/sharetribe/ftw-daily/pull/1399)
- [fix] Make sure that the verify email API endpoint has been called successfully before redirecting
  the user away from EmailVerificationPage.
  [#1397](https://github.com/sharetribe/ftw-daily/pull/1397)

  [v9.3.0]: https://github.com/sharetribe/ftw-hourly/compare/v9.2.0...v9.3.0

## [v9.2.0] 2020-12-16

### Updates from upstream (FTW-daily v7.2.0)

- [add] Add helper functions for setting up your own OIDC authentication and using FTW server as
  proxy when needed. [#1383](https://github.com/sharetribe/ftw-daily/pull/1383)

[v9.2.0]: https://github.com/sharetribe/ftw-hourly/compare/v9.1.0...v9.2.0

## [v9.1.0] 2020-12-15

- [fix] Remove deprecated time zone keys that Node implementation doesn't recongnize anymore.
  [#124](https://github.com/sharetribe/ftw-hourly/pull/124)

### Updates from upstream (FTW-daily v7.1.0)

- [change] Handle entity update with sparse attributes.
  [#1392](https://github.com/sharetribe/ftw-daily/pull/1392)
- [change] Remove react-google-maps dependency. It has not been maintained for 3 years. From now on,
  we use Google Maps API directly. However, the default map provider is still Mapbox.
  [#1389](https://github.com/sharetribe/ftw-daily/pull/1389)
- [fix] Pass metadata through sanitizeUser function.
  [#1391](https://github.com/sharetribe/ftw-daily/pull/1391)
- [fix] Call for the same page caused unnecessary rendering
  [#1388](https://github.com/sharetribe/ftw-daily/pull/1388)
- [fix] Fix Google Maps default centering if no bounds or center is given.

src/config.js
import * as custom from './marketplace-custom-config.js';
import defaultLocationSearches from './default-location-searches';
import { defaultMCC, stripePublishableKey, stripeCountryDetails } from './stripe-config';
import { currencyConfiguration } from './currency-config';

const env = process.env.REACT_APP_ENV;
const dev = process.env.REACT_APP_ENV === 'development';

// If you want to change the language, remember to also change the
// locale data and the messages in the app.js file.
const locale = 'en';
const i18n = {
  /*
    0: Sunday
    1: Monday
    ...
    6: Saturday
  */
  firstDayOfWeek: 0,
};

// Should search results be ordered by distance to origin.
// NOTE: If this is set to true add parameter 'origin' to every location in default-location-searches.js
// Without the 'origin' parameter, search will not work correctly
// NOTE: Keyword search and ordering search results by distance can't be used at the same time. You can turn keyword
// search off by changing the keywordFilterConfig parameter active to false in marketplace-custom-config.js
const sortSearchByDistance = false;

// API supports custom processes to be used in booking process.
// We need to specify it when we are initiating a new order
// (or fetching price information by calling 'initiateSpeculative' endpoint).
//
// In a way, 'processAlias' defines which transaction process (or processes)
// this particular web application is able to handle.
const bookingProcessAlias = 'flex-hourly-default-process/release-1';

// The transaction line item code for the main unit type in bookings.
//
// Possible values: ['line-item/night', 'line-item/day', 'line-item/units';]
//
// Note 1: This 'bookingUnitType' variable affects only web app.
//         If you are using privileged transitions (which is used by the default process),
//         you also need to configure unit type in API server: server/api-util/lineItems.js
//
// Note 2: Translations will use different translation keys for night, day or unit
//         depending on the value chosen.
const bookingUnitType = 'line-item/units';

// Should the application fetch available time slots (currently defined as
// start and end dates) to be shown on listing page.
const enableAvailability = process.env.REACT_APP_AVAILABILITY_ENABLED === 'true';

// A maximum number of days forwards during which a booking can be made.
// This is limited due to Stripe holding funds up to 90 days from the
// moment they are charged:
// https://stripe.com/docs/connect/account-balances#holding-funds
//
// See also the API reference for querying time slots:
// https://www.sharetribe.com/api-reference/marketplace.html#query-time-slots
const dayCountAvailableForBooking = 90;

// To pass environment variables to the client app in the build
// script, react-scripts (and the sharetribe-scripts fork of
// react-scripts) require using the REACT_APP_ prefix to avoid
// exposing server secrets to the client side.
const sdkClientId = process.env.REACT_APP_SHARETRIBE_SDK_CLIENT_ID;
const sdkBaseUrl = process.env.REACT_APP_SHARETRIBE_SDK_BASE_URL;
const sdkTransitVerbose = process.env.REACT_APP_SHARETRIBE_SDK_TRANSIT_VERBOSE === 'true';

// Marketplace currency.
// It should match one of the currencies listed in currency-config.js
const currencyConf = process.env.REACT_APP_SHARETRIBE_MARKETPLACE_CURRENCY;
const currency = currencyConf ? currencyConf.toUpperCase() : currencyConf;

// Currency formatting options.
// See: https://github.com/yahoo/react-intl/wiki/API#formatnumber
const currencyConfig = currencyConfiguration(currency);

// Listing minimum price in currency sub units, e.g. cents.
// 0 means no restriction to the price
const listingMinimumPriceSubUnits = 0;

// Sentry DSN (Data Source Name), a client key for authenticating calls to Sentry
const sentryDsn = process.env.REACT_APP_SENTRY_DSN;

// If webapp is using SSL (i.e. it's behind 'https' protocol)
const usingSSL = process.env.REACT_APP_SHARETRIBE_USING_SSL === 'true';

// Address information is used in SEO schema for Organization (http://schema.org/PostalAddress)
const addressCountry = 'FI';
const addressRegion = 'Helsinki';
const postalCode = '00100';
const streetAddress = 'Bulevardi 14';

// Canonical root url is needed in social media sharing and SEO optimization purposes.
const canonicalRootURL = process.env.REACT_APP_CANONICAL_ROOT_URL;

// Site title is needed in meta tags (bots and social media sharing reads those)
const siteTitle = 'Culinary Chefz';

// Twitter handle is needed in meta tags (twitter:site). Start it with '@' character
const siteTwitterHandle = '@sharetribe';

// Instagram page is used in SEO schema (http://schema.org/Organization)
const siteInstagramPage = null;

// Facebook page is used in SEO schema (http://schema.org/Organization)
const siteFacebookPage = 'https://www.facebook.com/Sharetribe/';

// Social logins & SSO

// Note: Facebook app id is also used for tracking:
// Facebook counts shares with app or page associated by this id
// Currently it is unset, but you can read more about fb:app_id from
// https://developers.facebook.com/docs/sharing/webmasters#basic
// You should create one to track social sharing in Facebook
const facebookAppId = process.env.REACT_APP_FACEBOOK_APP_ID;

const maps = {
  mapboxAccessToken: process.env.REACT_APP_MAPBOX_ACCESS_TOKEN,
  googleMapsAPIKey: process.env.REACT_APP_GOOGLE_MAPS_API_KEY,

  // The location search input can be configured to show default
  // searches when the user focuses on the input and hasn't yet typed
  // anything. This reduces typing and avoids too many Geolocation API
  // calls for common searches.
  search: {
    // When enabled, the first suggestion is "Current location" that
    // uses the browser Geolocation API to query the user's current
    // location.
    suggestCurrentLocation: process.env.REACT_APP_DEFAULT_SEARCHES_ENABLED === 'true',

    // Distance in meters for calculating the bounding box around the
    // current location.
    currentLocationBoundsDistance: 1000,

    // Example location can be edited in the
    // `default-location-searches.js` file.
    defaults:
      process.env.REACT_APP_DEFAULT_SEARCHES_ENABLED === 'true' ? defaultLocationSearches : [],

    // Limit location autocomplete to a one or more countries
    // using ISO 3166 alpha 2 country codes separated by commas.
    // If you want to limit the autocomplete, uncomment this value:
    // countryLimit: ['AU'],
  },

  // When fuzzy locations are enabled, coordinates on maps are
  // obfuscated randomly around the actual location.
  //
  // NOTE: This only hides the locations in the UI level, the actual
  // coordinates are still accessible in the HTTP requests and the
  // Redux store.
  fuzzy: {
    enabled: false,

    // Amount of maximum offset in meters that is applied to obfuscate
    // the original coordinates. The actual value is random, but the
    // obfuscated coordinates are withing a circle that has the same
    // radius as the offset.
    offset: 500,

    // Default zoom level when showing a single circle on a Map. Should
    // be small enough so the whole circle fits in.
    defaultZoomLevel: 13,

    // Color of the circle on the Map component.
    circleColor: '#c0392b',
  },

  // Custom marker image to use in the Map component.
  //
  // NOTE: Not used if fuzzy locations are enabled.
  customMarker: {
    enabled: false,

    // Publicly accessible URL for the custom marker image.
    //
    // The easiest place is /public/static/icons/ folder, but then the
    // marker image is not available while developing through
    // localhost.
    url: encodeURI(`${canonicalRootURL}/static/icons/map-marker-32x32.png`),

    // Dimensions of the marker image.
    width: 32,
    height: 32,

    // Position to anchor the image in relation to the coordinates,
    // ignored when using Mapbox.
    anchorX: 16,
    anchorY: 32,
  },
};

// NOTE: only expose configuration that should be visible in the
// client side, don't add any server secrets in this file.
const config = {
  env,
  dev,
  locale,
  bookingProcessAlias,
  bookingUnitType,
  enableAvailability,
  dayCountAvailableForBooking,
  i18n,
  sdk: {
    clientId: sdkClientId,
    baseUrl: sdkBaseUrl,
    transitVerbose: sdkTransitVerbose,
  },
  sortSearchByDistance,
  currency,
  currencyConfig,
  listingMinimumPriceSubUnits,
  stripe: {
    defaultMCC: defaultMCC,
    publishableKey: stripePublishableKey,
    supportedCountries: stripeCountryDetails,
  },
  canonicalRootURL,
  address: {
    addressCountry,
    addressRegion,
    postalCode,
    streetAddress,
  },
  siteTitle,
  siteFacebookPage,
  siteInstagramPage,
  siteTwitterHandle,
  facebookAppId,
  sentryDsn,
  usingSSL,
  maps,
  custom,
};

export default config;
